src/core/dnd/types.ts

```diff
@@ -59,76 +58,94 @@ export const canDropIntoLayout = (
 ) => {
   // check scope changes
   const detailContainer = getDetailContainer(layout);
-  if (!canDropIntoScope(item, detailContainer)) {
-    return false;
-  }
-  // check whether
-  return true;
+  return canDropIntoScope(item, detailContainer);
 };
 
 /**
  * Check whether the element to drop fits into the given scope,
  * e.g. whether a nested array object is dropped into the correct array ui schema control.
  *
  * @param item the drag and drop item
- * @param scopeChangingElement the nearest scope changing element,
+ * @param scopeUISchemaElement the nearest scope changing element,
  * e.g. the nearest array control into which shall be dropped.
  * Use `undefined` when dropping outside of any scope changing element.
  */
 export const canDropIntoScope = (
   item: NewUISchemaElement,
   scopeUISchemaElement: EditorUISchemaElement | undefined
 ) => {
-  // it's a control when there is a schema
-  if (item.schema) {
-    // check wether the control fits to the scope
-    // -- TODO check other cases than array
-    const scopeSchemaElement = getArrayContainer(item.schema);
-    if (!!scopeSchemaElement !== !!scopeUISchemaElement) {
-      // scopes don't match when one exist and the other doesn't
-      return false;
-    }
-    if (
-      scopeSchemaElement &&
-      scopeUISchemaElement &&
-      scopeUISchemaElement.linkedSchemaElement !== scopeSchemaElement.uuid
-    ) {
-      // scopes didn't match
+  const controlObject = item.schema;
+  if (controlObject) {
+    const scopeSchemaElement = getScopeChangingContainer(controlObject);
+    if (!scopesMatch(scopeSchemaElement, scopeUISchemaElement)) {
       return false;
     }
   }
   return true;
 };
 
-export const canMoveSchemaElementTo = (
-  item: MoveUISchemaElement,
-  layout: EditorUISchemaElement,
-  index: number
+/**
+ * Scopes match if they are linked or both don't exist.
+ */
+const scopesMatch = (
+  schemaScope: SchemaElement | undefined,
+  uiScope: EditorUISchemaElement | undefined
 ) => {
-  const uiElementToMove = item.uiSchemaElement as EditorUISchemaElement;
-  // can't move the root element
-  if (!uiElementToMove.parent) {
-    return false;
+  if (!schemaScope && !uiScope) {
+    return true;
   }
-  // can't move element into itself
-  if (getHierarchy(layout).includes(uiElementToMove)) {
+  if (!(schemaScope && uiScope)) {
     return false;
   }
-  // can't move element next to itself (which would result in no change)
-  if (layout === uiElementToMove.parent) {
-    const currentIndex = (uiElementToMove.parent as EditorLayout).elements.indexOf(
-      uiElementToMove
+  return uiScope.linkedSchemaElement === schemaScope.uuid;
```

Comment:
wouldn't `uiScope?.linkedSchemaElement === schemaScope?.uuid;` be enough or would we miss a case?

Reply:
In that case we would return `true` when `schemaScope` is undefined but the `uiScope` exists without a linked schema element, e.g. we would allow dropping `birthDate` into the array control although the property doesn't exist there.

Reply:
Now thinking about it, the `uiScope` should always have a `linkedSchemaElement` in our case when it exists. So something is going wrong here and your code suggestion should be sufficient.

Reply:
Ah okay, the `linkedSchemaElement` is only not set when the form is nothing but the array control, and this occurs because of #57. I'll simplify the code. Good catch!